### PR TITLE
[feature/fileprovider-actions] Add Actions to File Provider: Sharing & Public Links

### PR DESCRIPTION
**Pull request by hosy**
## Description
This PR adds two custom actions to the File Provider: Sharing and Public Links, which will open the UI for adding and editing sharing and public links to the selected item directly from the file provider.

## Motivation and Context
Provide this core features from any app which is using the file provider.

## How Has This Been Tested?
- Open file provider from Files.app our any third party app
- Long press an item
- Select `Sharing` or `Public Links`

## Screenshots (if appropriate):

| Simulator Screen Shot - iPad Pro (11-inch) (2nd generation) - 2021-02-26 at 09 10 02 | Simulator Screen Shot - iPad Pro (11-inch) (2nd generation) - 2021-02-26 at 12 07 07 |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Simulator Screen Shot - iPad Pro (11-inch) (2nd generation) - 2021-02-26 at 12 10 12 | Simulator Screen Shot - iPad Pro (11-inch) (2nd generation) - 2021-02-26 at 09 11 53 |

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] Added changelog files for the fixed issues in folder changelog/unreleased

## QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%2011.6/Sharing%20Files%20%20app.md

- [X] (1) Connecting screen happens sometimes https://github.com/owncloud/ios-app/pull/910#issuecomment-804897547 [FIXED]
- [x] (2) Dark theme everywhere https://github.com/owncloud/ios-app/pull/910#issuecomment-804898456 [FIXED]
- [x] (3) Stucked after unsharing https://github.com/owncloud/ios-app/pull/910#issuecomment-805741742 [FIXED]
- [x] (4) Sharing aPI disabled behaviour https://github.com/owncloud/ios-app/pull/910#issuecomment-805746680 [FIXED]

### CODE REVIEW

File: ownCloud File Provider UI/Info.plist [hunk 1]
```diff
@@ -23,7 +23,24 @@
...
 		<key>NSExtensionFileProviderActions</key>
-		<array/>
+		<array>
+			<dict>
+				<key>NSExtensionFileProviderActionActivationRule</key>
+				<string>TRUEPREDICATE</string>
+				<key>NSExtensionFileProviderActionIdentifier</key>
+				<string>com.owncloud.FileProviderUI.Share</string>
+				<key>NSExtensionFileProviderActionName</key>
+				<string>Share this Item</string>
```
Comment (felix-schwarz) [inline ownCloud File Provider UI/Info.plist:33]: As discussed in the coordination call earlier, this should be renamed to `Share with user/group`.

File: ownCloud File Provider UI/Info.plist [hunk 2]
```diff
@@ -23,7 +23,24 @@
...
+			</dict>
+			<dict>
+				<key>NSExtensionFileProviderActionActivationRule</key>
+				<string>TRUEPREDICATE</string>
+				<key>NSExtensionFileProviderActionIdentifier</key>
+				<string>com.owncloud.FileProviderUI.PublicLinks</string>
+				<key>NSExtensionFileProviderActionName</key>
+				<string>Links</string>
```
Comment (felix-schwarz) [inline ownCloud File Provider UI/Info.plist:41]: As discussed in the coordination call earlier, this should be renamed to `Share link`.

File: ownCloud File Provider UI/DocumentActionViewController.swift [hunk 3]
```diff
@@ -20,11 +20,101 @@ import UIKit
...
+			addChild(themeNavigationController)
+		}
+
+		showCancelLabel(with: "Connecting…".localized)
+
+		var actionTypeLabel = ""
+		var actionExtensionType : ActionExtensionType = .undefined
+		if actionIdentifier == "com.owncloud.FileProviderUI.Share" {
+			actionExtensionType = .sharing
+			actionTypeLabel = "Sharing".localized
```
Comment (felix-schwarz) [inline ownCloud File Provider UI/DocumentActionViewController.swift:67]: The action name, if changed, also needs to be changed here.

**Review by felix-schwarz — CHANGES_REQUESTED** [review]

File: ownCloud File Provider UI/Info.plist (hunk 1 above)
Comment (hosy) [reply ownCloud File Provider UI/Info.plist]: fixed!

File: ownCloud File Provider UI/Info.plist (hunk 2 above)
Comment (hosy) [reply ownCloud File Provider UI/Info.plist]: fixed!

File: ownCloud File Provider UI/CancelLabelViewController.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import FileProviderUI
+import ownCloudApp
+import ownCloudAppShared
+
+class CancelLabelViewController: UIViewController {
+
+	@IBOutlet var label : UILabel!
+	@IBOutlet var button : ThemeButton!
+
+    override func viewDidLoad() {
```
[the remark this answers is not in the document]
Comment (hosy) [reply ownCloud File Provider UI/CancelLabelViewController.swift]: removed

File: ownCloud File Provider UI/DocumentActionViewController.swift (hunk 3 above)
Comment (hosy) [reply ownCloud File Provider UI/DocumentActionViewController.swift]: adopted

File: ownCloud File Provider UI/DocumentActionViewController.swift [hunk 5]
```diff
@@ -20,11 +20,101 @@ import UIKit
...
+		} else if actionIdentifier == "com.owncloud.FileProviderUI.PublicLinks" {
+			actionExtensionType = .links
+			actionTypeLabel = "Public Links".localized
```
[the remark this answers is not in the document]
Comment (hosy) [reply ownCloud File Provider UI/DocumentActionViewController.swift]: adopted

File: ownCloud File Provider UI/DocumentActionViewController.swift [hunk 6]
```diff
@@ -20,11 +20,101 @@ import UIKit
...
+							} else if actionExtensionType == .links {
+								let publicLinkController = PublicLinkTableViewController(core: core, item: item)
+								publicLinkController.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(self.dismissView))
+								self.themeNavigationController?.viewControllers = [publicLinkController]
+							}
+						} else if core.connectionStatus == .connecting {
+							triedConnecting = true
+							self.showCancelLabel(with: "Connecting…".localized)
+						} else if core.connectionStatus == .offline || core.connectionStatus == .unavailable {
+							if triedConnecting {
```
[the remark this answers is not in the document]
Comment (hosy) [reply ownCloud File Provider UI/DocumentActionViewController.swift:113]: used your provided code

**Review by felix-schwarz — APPROVED** [review]

**Comment by felix-schwarz** [comment]
Thanks for the changes! Noticed after reviewing and approving the changes that there are merge conflicts that you might want to look into.

**Comment by jesmrec** [comment]
### (1)

This is not something that happens every time, but i can reproduce easily after selecting the new Sharing options. The view stucks in `Connecting`. The only way is cancelling and trying again.

https://user-images.githubusercontent.com/14894746/112152648-ce016600-8be2-11eb-9408-dac955040eef.MP4

iPhoneXR v14.4 ` 8bad1fb`

**Comment by jesmrec** [comment]
### (2)

Dark theme is set by default in the sharing view inside Files App, no matter which theme is selected in the app.

Is this something fixable?

**Comment by hosy** [comment]
@jesmrec (2) should be fixed.

**Comment by jesmrec** [comment]
(2) is fixed

**Comment by hosy** [comment]
@jesmrec you can check (1) too.

**Comment by jesmrec** [comment]
### (3)

1. Alice share a folder with Bob
2. Bob opens Files App
3. Bob select the shared item and the option `Share with user/group`
4. Bob taps on `Unshare` 

Current: Spinner forever
Expected: Folder unshared

iPhoneXR v14.4 
`0f7b667`

**Comment by jesmrec** [comment]
(1) is fixed

**Comment by jesmrec** [comment]
### (4)

1. In webUI, go to admin dashboard -> Sharing, and disable `Allow apps to use the Share API`.  From now on, no sharing with users and public link creation will be allowed.
2. Open Files App, ownCloud account

Current:

Sharing with users/groups: it is enabled and it is allowed to search a recipient. Finally, you get the proper error because API is not enabled.
Share Link: Not allowed to create from scratch

Expected: 

Both `Share with users/groups` and `Share link` should behave in the same way: both allowing to create and then returning an error, or both disabled from scratch preventing users to do actions that will finish with an error.

In the app, both are disabled when Sharing API is disabled, as reference.

iPhoneXR v14.4
`0f7b667`

**Comment by hosy** [comment]
@jesmrec (3), (4) is fixed now. Regarding (4): it is not possible to hide the menu item, instead a message will be shown, that sharing is not allowed.

**Comment by jesmrec** [comment]
> @jesmrec (3), (4) is fixed now. Regarding (4): it is not possible to hide the menu item, instead a message will be shown, that sharing is not allowed.

yes, i know. The option is there and then, the message inside. This is totally correct.

(3) and (4) are fixed and test plan execution is completed

Approved

**Review by jesmrec — APPROVED** [review]

**Pull request merged**